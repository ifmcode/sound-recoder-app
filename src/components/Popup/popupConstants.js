const PopupType = {
  DELETE_POPUP: "delete-popup",
  EDIT_POPUP: "edit-popup",
}

const PopupConfig = {
  [PopupType.DELETE_POPUP]: {
    title: "Are you sure you want to delete this recording?",
    buttons: [{
      value: "Delete",
      confirmAction: true,
    }, {
      value: "Cancel",
      confirmAction: false,
    }],
  },
  [PopupType.EDIT_POPUP]: {
    title: "Enter a new name",
    buttons: [{
      value: "Edit",
      confirmAction: true,
    }, {
      value: "Cancel",
      confirmAction: false,
    }],
  },
}

export {
  PopupConfig,
  PopupType,
}
