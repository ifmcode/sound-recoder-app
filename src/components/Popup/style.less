@import (reference) "../../styles/colors.less";
@import (reference) "../../styles/sizes.less";

.popup {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  box-sizing: border-box;
  z-index: 3;
  .text-container {
    position: absolute;
    left: 10%;
    top: 35%;
    transform: translateY(-50%);
    width: 80%;
    height: auto;
    background-color: @grey;
    border: 3px solid @bluegreen;
    border-radius: 6px;
    padding: 14px;
    box-sizing: border-box;
    text-align: center;

    .popup-title {
      position: relative;
      font-size: @font-size_18;
      color: @white;
      margin-bottom: 30px;
      display: block;
      text-align: center;
    }
    
    .popup-advice {
      position: relative;
      font-size: @font-size_18;
      color: @bluegreen;
      margin-bottom: 8px;
    }

    .popup-input {
      margin-bottom: 25px;
      width: 220px;
      height: 30px;
      padding: 0 10px;
      box-sizing: border-box;
      border-radius: 6px;

      &:focus{
        outline: none;
      }
    }

    .buttons-container {
      position: relative;
      display: flex;
      justify-content: space-around;
      margin-bottom: 10px;
      .popup-button {
        position: relative;
        padding: 3px 20px;
        font-size: @font-size_18;
        color: @bluegreen;
        background-color: @darkgrey;
        cursor: pointer;
        border: 2px solid @black;
        transition: all .5s ease-in-out;
        &:hover {
          background-color: @black;
        }
      }
    }
  }
}