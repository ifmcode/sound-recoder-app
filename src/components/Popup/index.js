import "./style.less";
import React, {useEffect, useRef, useState} from 'react';
import {PopupConfig, PopupType} from './popupConstants';

/**
 * 
 * @param {PopupType} popupType // It could be 'delete-popup' or 'edit-popup'
 * @param {Object} selectedRecording // This is the recording that is being manipulated
 * @param {Function} onConfirm // Funtion that it call when confirm button is clicked
 * @param {Function} onDeny // Funtion that it call when cancel button is clicked
 * @returns 
 */
const Popup = ({popupType, selectedRecording, adviceMessage, onConfirm, onDeny}) => {

  const inputName = useRef();

  return (
    <div className="popup">
      <div className='text-container'>
        <label className='popup-title'>{PopupConfig[popupType]?.title}</label>
        <label className="popup-advice">{adviceMessage}</label>
        {
          (popupType === PopupType.EDIT_POPUP) &&
          <input
            className="popup-input"
            type="text"
            ref={inputName}
            placeholder={selectedRecording.name}/>
        }
        <div className='buttons-container'>
          {
            PopupConfig[popupType]?.buttons.map((button, index) => {
              return (
                <button 
                  key= {index}
                  className='popup-button'
                  onClick={() => button.confirmAction ? onConfirm(inputName?.current?.value || "") : onDeny()}>
                  {button.value}
                </button>
              );
            })
          }
        </div>
      </div>
    </div>
  );
};

export default Popup;
