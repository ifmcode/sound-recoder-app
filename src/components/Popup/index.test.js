import React from 'react';
import {render, screen, fireEvent} from '@testing-library/react'
import Popup from '.';
import {PopupType} from './popupConstants';

/**
 * Following the Object mother pattern we have this small fn that generates a valid object
 * that matches the structure of a recording
 *
 * @param {{ idNumber?: number, name?: string }} options
 * @returns {{ stream: string, name: string, id: string }}
 */
function getMockRecording({
  idNumber = Math.floor(Math.random() * 100),
  name = new Date().toISOString().split('.')[0].split('T').join(' ') + idNumber
}) {
  return {
    stream: "audioUrl",
    name,
    id: `id${idNumber}`
  }
}

describe('Testing Popup component when is used as a delete popup', () => {

  let wrapper;
  const selectedRecording = getMockRecording({}),
    onConfirm = jest.fn(),
    onDeny = jest.fn();

  beforeEach(() => {
    wrapper = render(
      <Popup
        popupType={PopupType.DELETE_POPUP}
        selectedRecording={selectedRecording}
        onConfirm={onConfirm}
        onDeny={onDeny}
      />
    );
  });
  
  test("should render the Popup component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('renders the deleting popup title', () => {
    expect(screen.queryByText('Are you sure you want to delete this recording?')).toBeInTheDocument();
  });
  
  test('renders delete and cancel button', () => {
    expect(screen.getByRole('button', {name: 'Delete'})).toBeInTheDocument();
    expect(screen.getByRole('button', {name: 'Cancel'})).toBeInTheDocument();
  });

  test("should call the onConfirm function when click in the delete button", () => {
    const deleteButton = screen.getByRole('button', {name: 'Delete'});
    fireEvent.click(deleteButton);
    expect(onConfirm).toHaveBeenCalledTimes(1);
  });
  
  test("should call the onDeny function when click in the cancel button", () => {
    const cancelButton = screen.getByRole('button', {name: 'Cancel'});
    fireEvent.click(cancelButton);
    expect(onDeny).toHaveBeenCalledTimes(1);
  });
});

describe('Testing Popup component when is used as a edit popup', () => {

  let wrapper;
  const selectedRecording = getMockRecording({}),
    onConfirm = jest.fn(),
    onDeny = jest.fn();

  beforeEach(() => {
    wrapper = render(
      <Popup
        popupType={PopupType.EDIT_POPUP}
        selectedRecording={selectedRecording}
        onConfirm={onConfirm}
        onDeny={onDeny}
      />
    );
  });
  
  test("should render the Popup component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('renders the editing popup title', () => {
    expect(screen.queryByText('Enter a new name')).toBeInTheDocument();
  });
  
  test('renders the editing popup input', () => {
    expect(screen.getByRole('textbox')).toBeInTheDocument();
  });
  
  test('renders edit and cancel button', () => {
    expect(screen.getByRole('button', {name: 'Edit'})).toBeInTheDocument();
    expect(screen.getByRole('button', {name: 'Cancel'})).toBeInTheDocument();
  });

  test("should call the onConfirm function when click in the edit button", () => {
    const editButton = screen.getByRole('button', {name: 'Edit'});
    fireEvent.click(editButton);
    expect(onConfirm).toHaveBeenCalledTimes(1);
  });
  
  test("should call the onDeny function when click in the cancel button", () => {
    const cancelButton = screen.getByRole('button', {name: 'Cancel'});
    fireEvent.click(cancelButton);
    expect(onDeny).toHaveBeenCalledTimes(1);
  });
})
