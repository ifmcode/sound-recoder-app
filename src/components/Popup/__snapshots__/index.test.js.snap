// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Popup component when is used as a delete popup should render the Popup component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="popup"
      >
        <div
          class="text-container"
        >
          <label
            class="popup-title"
          >
            Are you sure you want to delete this recording?
          </label>
          <div
            class="buttons-container"
          >
            <button
              class="popup-button"
            >
              Delete
            </button>
            <button
              class="popup-button"
            >
              Cancel
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="popup"
    >
      <div
        class="text-container"
      >
        <label
          class="popup-title"
        >
          Are you sure you want to delete this recording?
        </label>
        <div
          class="buttons-container"
        >
          <button
            class="popup-button"
          >
            Delete
          </button>
          <button
            class="popup-button"
          >
            Cancel
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Testing Popup component when is used as a edit popup should render the Popup component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="popup"
      >
        <div
          class="text-container"
        >
          <label
            class="popup-title"
          >
            Enter a new name
          </label>
          <input
            class="popup-input"
            placeholder="2022-01-26 17:04:082"
            type="text"
          />
          <div
            class="buttons-container"
          >
            <button
              class="popup-button"
            >
              Edit
            </button>
            <button
              class="popup-button"
            >
              Cancel
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="popup"
    >
      <div
        class="text-container"
      >
        <label
          class="popup-title"
        >
          Enter a new name
        </label>
        <input
          class="popup-input"
          placeholder="2022-01-26 17:04:082"
          type="text"
        />
        <div
          class="buttons-container"
        >
          <button
            class="popup-button"
          >
            Edit
          </button>
          <button
            class="popup-button"
          >
            Cancel
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
